zapcore/sampler.go

```diff
@@ -62,33 +69,85 @@ func fnv32a(s string) uint32 {
 	return hash
 }
 
-func (c *counter) IncCheckReset(t time.Time, tick time.Duration) uint64 {
+func (c *counter) IncCheckReset(t time.Time, tick time.Duration) (currentVal, preResetVal uint64) {
 	tn := t.UnixNano()
 	resetAfter := c.resetAt.Load()
 	if resetAfter > tn {
-		return c.counter.Inc()
+		return c.counter.Inc(), 0
 	}
 
-	c.counter.Store(1)
-
 	newResetAfter := tn + tick.Nanoseconds()
 	if !c.resetAt.CAS(resetAfter, newResetAfter) {
-		// We raced with another goroutine trying to reset, and it also reset
-		// the counter to 1, so we need to reincrement the counter.
-		return c.counter.Inc()
+		// We raced with another goroutine trying to reset, and it reset the
+		// counter to 1, so we just increment.
+		return c.counter.Inc(), 0
 	}
+	preResetVal = c.counter.Load()
+	currentVal = 1
+	c.counter.Store(currentVal)
```

Comment:
i think this change can now "lose" increments

previously, we assumed that the `resetAt` was changed *after* we set the counter value. now it's set after, so we could have:

g1 changes `resetAt`
g2 tries to change `resetAt`, fails, and then increments counter (let's say to 5)
g1 now stores `1` in the counter

we've lost g2's increment.